File: Algorithms/Sorting-Algorithm/Tim_Sort/Tim_Sort.py
minrun = 32

def InsSort(arr,start,end):    
    for i in range(start+1,end+1):
        elem = arr[i]
        j = i-1
        while j>=start and elem<arr[j]:
            arr[j+1] = arr[j]
            j -= 1
        arr[j+1] = elem
    return arr

def merge(arr,start,mid,end):
    if mid==end:
        return arr
    first = arr[start:mid+1]
    last = arr[mid+1:end+1]
    len1 = mid-start+1
    len2 = end-mid
    ind1 = 0
    ind2 = 0
    ind  = start
     
    while ind1<len1 and ind2<len2:
        if first[ind1]<last[ind2]:
            arr[ind] = first[ind1]
            ind1 += 1
        else:
            arr[ind] = last[ind2]
            ind2 += 1
        ind += 1
     
    while ind1<len1:
        arr[ind] = first[ind1]
        ind1 += 1
        ind += 1
              
    while ind2<len2:
        arr[ind] = last[ind2]
        ind2 += 1
        ind += 1   
              
    return arr
            

def TimSort(arr):
    n = len(arr)
    
    for start in range(0,n,minrun):
        end = min(start+minrun-1,n-1)
        arr = InsSort(arr,start,end)
        
    curr_size = minrun
    while curr_size<n:    
        for start in range(0,n,curr_size*2):
            mid = min(n-1,start+curr_size-1)
            end = min(n-1,mid+curr_size)
            arr = merge(arr,start,mid,end)
        curr_size *= 2
    return arr

File: Algorithms/Sorting-Algorithm/Tim_Sort/Tim_Sort.c
//C Program for Tim Sort
#include<stdio.h>

const int run = 32;
//Function to find minimum of two numbers
int min (int a, int b)
{
  if (a < b)
    return a;
  else
    return b;
}
//function implementing insertion sort 
void insertionSort (int a[], int low, int high)
{
  int i, j, temp;
  for (i = low + 1; i <= high; i++)
    {
      temp = a[i];
      j = i - 1;
      while (a[j] > temp && j >= low)
        a[j + 1] = a[j--];
      a[j + 1] = temp;
    }
}
//function implementing merge sort
void merge (int a[], int left, int mid, int right)
{
  int len1 = mid - left + 1;
  int len2 = right - mid;
  int low[len1], high[len2];
  int i, j, k;
  for (i = 0; i < len1; i++)
    low[i] = a[left + i];
  for (i = 0; i < len2; i++)
    high[i] = a[mid + 1 + i];

  i = 0; j = 0;
  k = left;

  while (i < len1 && j < len2)
    {
      if (low[i] <= high[j])
	   a[k] = low[i++];
      else
	   a[k] = high[j++];
      k++;
    }
    while (i < len1)
      a[k++] = low[i++];

    while (j < len2)
      a[k++] = high[j++];
}

void timSort (int a[], int n)
{
  int i, size, low, mid, high;
  for (i = 0; i < n; i += run)
    insertionSort (a, i, min ((i + 31), (n - 1)));
  for (size = run; size < n; size = 2 * size)
  {
    for (low = 0; low < n; low += 2 * size)
	{
	  mid = low + size - 1;
	  high = min ((low + 2 * size - 1), (n - 1));
      merge (a, low, mid, high);
	}
  }
}

void main ()
{
  int a[50] , n, i;
  //Get number of elements from user
  printf("Enter number of elements to be sorted:");
  scanf("%d",&n);
  //Get elements from user into array
  printf("Enter the elements:");
  for(i=0;i<n;i++)
    scanf("%d",&a[i]);
  //Apply timsort to the array
  timSort (a, n);
  //Print the sorted array
  printf ("Printing sorted array elements \n");
  for (i = 0; i < n; i++)
    printf ("%d  ", a[i]);
  printf ("\n");
}
